torch/csrc/jit/passes/xnnpack_rewrite.cpp

```diff
@@ -81,8 +81,105 @@ void insertPrePackedConv2dOp(std::shared_ptr<Graph>& graph) {
         return (%r) )";
 
   SubgraphRewriter rewriter;
-  rewriter.RegisterRewritePattern(
-      conv_2d_pattern, prepacked_ops_conv2d_pattern);
+  rewriter.RegisterRewritePattern(conv_2d_pattern, prepacked_ops_conv2d_pattern);
+  rewriter.runOnGraph(graph);
+}
+
+bool isReluFusable(const std::shared_ptr<Graph>& graph) {
+  using graph_rewrite_helper::PatternInfo;
+  const PatternInfo& pattern_relu = PatternInfo::parse_from_str(R"(
+      graph(%input):
+        %r = aten::relu(%input)
+        return (%r) )");
+  const Graph& pattern_relu_graph = *pattern_relu.pattern_graph;
+  const auto& relu_vmap = pattern_relu.vmap;
+  const auto& matches = findPatternMatches(pattern_relu_graph, *graph);
+  for (const auto& match : matches) {
+    auto relu_input_value = match.values_map.at(relu_vmap.at("input"));
+    // If the input to relu is used by something other than relu then such
+    // a relu cannot be fused.
+    if (relu_input_value->uses().size() > 1) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// Fusing Relu with Linear/Conv already implies that relu will
+// be done inplace. However, if the input of the relu was not to be
+// overwritten then it should not be fused.
+void fuseReluWithPackedOps(std::shared_ptr<Graph>& graph) {
+
+  SubgraphRewriter rewriter;
+
+  std::string linear_prepack_run_relu_fused = R"(
+    graph(%input, %weight, %bias, %output_min_max):
+        %output_min: float = prim::Constant[value=0.0]()
+        %output_max: float = prim::Constant[value=1.0]()
+        %packed_weight_bias = prepacked::linear_clamp_prepack(
+            %weight, %bias, %output_min, %output_max)
+        %res = prepacked::linear_clamp_run(%input, %packed_weight_bias)
+        return (%res))";
+
+  std::string conv2d_prepack_run_relu_fused = R"(
+    graph(%input, %weight, %bias, %stride:int[], %padding:int[],
+          %dilation:int[], %groups:int, %output_min_max):
+        %output_min: float = prim::Constant[value=0.0]()
+        %output_max: float = prim::Constant[value=1.0]()
+        %packed_weight_bias = prepacked::conv2d_clamp_prepack(
+            %weight, %bias, %stride, %padding, %dilation, %groups,
+            %output_min, %output_max)
+        %r = prepacked::conv2d_clamp_run(%input, %packed_weight_bias)
+        return (%r) )";
+
+  if (isReluFusable(graph)) {
```

Comment:
looks like this is checking for all `aten::relu` in the graph but the pattern is based on individual aten::relu?

I think when the input of relu is used once in the pattern it is only going to match the case when the input is used once as well, have you tried removing this check?

Reply:
> looks like this is checking for all `aten::relu` in the graph but the pattern is based on individual aten::relu?

Good point. Yes this checks for all relus.
> 
> I think when the input of relu is used once in the pattern it is only going to match the case when the input is used once as well, have you tried removing this check?

Oh I see. I need to remove this and add a test where output of conv is used elsewhere as well and check.

Reply:
btw, this is only true if the value is not input: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/ir/subgraph_matcher.cpp.html#100

Reply:
Ah. Makes sense. Actually, I dont need to test. This makes sense. I was only trying to check that relu is not fused with previous node, if the output of the previous node is used elsewhere. If input to relu is input to the pattern graph then there is nothing relu can fuse into.